### PR TITLE
test(sweeps): speed up tests 9x while maintaining code and conceptual coverage

### PR DESCRIPTION
**Pull request by kelu-wandb**
Previous tests were crazy slow and basically unnecessarily ran full end to end convergence tests. Refactored to target various edge cases.

- `linux310` test as of base PR:
  - Time: 22m 56s
  - Coverage: 91.8% https://app.codecov.io/gh/wandb/sweeps/pull/126/tree/src
- as of this PR:
  - Time: 2m 28s
  - Coverage: 93.6% https://app.codecov.io/gh/wandb/sweeps/pull/127/tree/src
  
Note that these tests don't just test code coverage, but also correctness of the algorithms themselves. They were just doing it by running massive convergence tests that took hundreds of iterations, which took forever. I tried to get it cut down by
- targeting a simple formula that converges quickly
- changing some tests to just check that things are getting better not necessarily fully converging
- cutting down on iterations in some cases, which seems good enough though might lead to slightly more flakes.
  - However, I specifically AVOIDED pinning a random seed and making them deterministic, because that could mean we just picked a lucky seed. So these tests will always be a little non-deterministic and slightly flaky, but hopefully not too much. (Even the old tests could flake occasionally.)

### CODE REVIEW

**Comment by kelu-wandb** [comment]
## Pull Request Stack
> [!WARNING]
> You may not want to merge this PR yet, because it is still stacked on another PR.

- `master`
  - #124
    - #125
      - #126
        - #127 ← *YOU ARE HERE*
          - #130

---
This PR stack is managed by [git-deck](http://github.com/wandb/git-deck)

File: tests/test_bayes_search.py [hunk 1]
```diff
@@ -23,19 +26,30 @@ def rosenbrock(x: ArrayLike) -> np.floating:
...
+    return (run.config["x"]["value"] - 0.3) ** 2
+
+
+def quadratic_run(x: floating) -> SweepRun:
+    run = SweepRun(config={"x": {"value": x}}, state=RunState.finished)
+    run.summary_metrics = {"loss": quadratic_loss(run)}
+    return run
+
+
+def simulate_bayes_search(
```
Comment (kmikowicz-wandb) [inline tests/test_bayes_search.py:43]: one nit is that there is an extremely small chance bayes picks 0.3 as its first point, so we don't actually test anything. might be worth restarting the simulation

**Review by kmikowicz-wandb — APPROVED** [review]
